Expression cycle detected

### Describe the bug

We have 2 variables in our pom.xml:
- git.build.version.tag (generated by git-commit-id-plugin with generateGitPropertiesFile and injectAllReactorProjects)
- image.tag referencing git.build.version.tag (like <image.tag>${git.build.version.tag}</image.tag>)

When using variable image.tag in our Dockerfile like ${image.tag}, kubernetes-maven-plugin gives the error "Expression cycle detected". There is no cycle though.

### Eclipse JKube version

1.9.1

### Component

Kubernetes Maven Plugin

### Apache Maven version

3.8.1

### Gradle version

None

### Steps to reproduce

1. use 2 variables in your pom.xml like 
- git.build.version.tag (generated by git-commit-id-plugin with generateGitPropertiesFile and injectAllReactorProjects)
- image.tag referencing git.build.version.tag (like <image.tag>${git.build.version.tag}</image.tag>)
2. use ${image.tag} variable in Dockerfile 
3. build the Dockerfile

### Expected behavior

image.tag should be resolved to the value of git.build.version.tag

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.25.3

### Environment

macOS

### Eclipse JKube Logs

```shell
Caused by: java.lang.IllegalArgumentException: Expression cycle detected, aborting.
    at org.eclipse.jkube.kit.common.JKubeFileInterpolator.checkPropertyWithDelimiters (JKubeFileInterpolator.java:137)
    at org.eclipse.jkube.kit.common.JKubeFileInterpolator.checkForPropertiesInLine (JKubeFileInterpolator.java:96)
    at org.eclipse.jkube.kit.common.JKubeFileInterpolator.interpolate (JKubeFileInterpolator.java:72)
    at org.eclipse.jkube.kit.common.JKubeFileInterpolator.interpolate (JKubeFileInterpolator.java:60)
    at org.eclipse.jkube.kit.build.api.helper.DockerFileUtil.extractLines (DockerFileUtil.java:111)
    at org.eclipse.jkube.kit.build.api.helper.DockerFileUtil.extractBaseImages (DockerFileUtil.java:67)
    at org.eclipse.jkube.kit.build.service.docker.BuildService.extractBaseFromDockerfile (BuildService.java:278)
    at org.eclipse.jkube.kit.build.service.docker.BuildService.autoPullBaseImage (BuildService.java:257)
    at org.eclipse.jkube.kit.build.service.docker.BuildService.buildImage (BuildService.java:75)
```


### Sample Reproducer Project

_No response_

### Additional context

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,6 +25,7 @@ Usage:
 * Fix #1273: Deprecate `jkube.io` annotation prefix in favor of `jkube.eclipse.org` for JKubeAnnotations
 * Fix #2079: Bump kubernetes-client to 6.5.0
 * Fix #2093: ClassCastException when extracting plugins from pom
+* Fix #2101: Resolve transitive properties
 
 ### 1.11.0 (2023-02-16)
 * Fix #1316: Add support for adding InitContainers via plugin configuration

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,7 +26,7 @@ Usage:
 * Fix #2086: Allow concurrent remote-dev sessions (service selector includes session id)
 * Fix #2093: ClassCastException when extracting plugins from pom
 * Fix #2104: Bump kubernetes-client to 6.5.1 (#2079)
-* Fix #2101: Resolve transitive properties
+* Fix #2101: Resolve dynamic transitive properties when creating Dockerfile
 
 ### 1.11.0 (2023-02-16)
 * Fix #1316: Add support for adding InitContainers via plugin configuration

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/JKubeFileInterpolator.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/JKubeFileInterpolator.java
@@ -22,6 +22,8 @@ import java.io.IOException;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
+import java.util.HashSet;
 
 public class JKubeFileInterpolator {
     public static final String DEFAULT_FILTER = "${*}";
@@ -137,24 +139,44 @@ public class JKubeFileInterpolator {
                     throw new IllegalArgumentException("Expression cycle detected, aborting.");
                 }
 
-                return replaceWithEscapeStr(line, searchPhrase, properties.get(property));
+                return replaceWithEscapeStr(line, searchPhrase, getPropertyValue(properties, expressionMarkers, property));
             }
         }
         return StringUtils.EMPTY;
     }
 
-    private static boolean isExpressionCycle(Map<String, String> properties, Map<String, String> expressionMarkers, String property) {
+    private static String getPropertyValue(Map<String, String> properties, Map<String, String> expressionMarkers, String property) {
         String value = properties.get(property);
-        // Check normal value
-        if (properties.get(value) != null) {
+        if (isProperty(value, expressionMarkers, properties)) {
+            return getPropertyValue(properties, expressionMarkers, parsePropertyKey(value, expressionMarkers));
+        } else {
+            return value;
+        }
+    }
+
+    private static boolean isExpressionCycle(Map<String, String> properties, Map<String, String> expressionMarkers, String property) {
+        Set<String> visited = new HashSet<>();
+        String value = properties.get(property);
+        while (isProperty(value, expressionMarkers, properties)) {
+            if (visited.contains(value)) {
+                return true;
+            }
+            visited.add(value);
+            value = properties.get(parsePropertyKey(value, expressionMarkers));
+        }
+        return false;
+    }
+
+    private static boolean isProperty(String property, Map<String, String> expressionMarkers, Map<String, String> properties) {
+        if (properties.containsKey(property)){
             return true;
         }
-
-        // Check value without delimiters
-        if (value != null) {
-            value = parsePropertyKey(value, expressionMarkers);
+        for (Map.Entry<String, String> entry : expressionMarkers.entrySet()) {
+            if (property.contains(entry.getKey())) {
+                return true;
+            }
         }
-        return properties.get(value) != null;
+        return false;
     }
 
     private static String parsePropertyKey(String property, Map<String, String> expressionMarkers) {

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/JKubeFileInterpolatorTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/JKubeFileInterpolatorTest.java
@@ -69,6 +69,19 @@ class JKubeFileInterpolatorTest {
     }
 
     @Test
+    void testShouldResolveTransitiveReferences() {
+        // Given
+        Properties props = new Properties();
+        props.setProperty("key1", "${key2}");
+        props.setProperty("key2", "${key3}");
+        props.setProperty("key3", "value");
+
+        String result = JKubeFileInterpolator.interpolate("${key1}", props, Collections.singletonMap("${", "}"));
+
+        assertThat(result).isEqualTo("value");
+    }
+
+    @Test
     @EnabledOnOs(OS.LINUX)
     void testShouldResolveByEnvVarInLinux() {
         // Given

